compiler/AST/AstToText.cpp

```diff
@@ -896,7 +898,139 @@ static bool looksLikeInfixOperator(const char *fnName)
   return !looksLikeIdentifier;
 }
 
-void AstToText::appendExpr(CallExpr* expr, bool printingType)
+/*
+ * Operator precedence according to the table in the spec,
+ * expressions.rst
+ *
+ * unary flag is needed because unary - (and +) have higher precedence
+ * than binary - (and +).
+ *
+ * postfix flag is needed because postfix ! has higher precedence than
+ * prefix !.
+ */
+static int opToPrecedence(const char *op, bool unary, bool postfix) {
+  // new is precedence 19, but doesn't come through this path.
+  if (postfix && (strcmp(op, "?") == 0 || strcmp(op, "!") == 0))
+    return 18;
+  else if (strcmp(op, ":") == 0)
+    return 17;
+  else if (strcmp(op, "**") == 0)
+    return 16;
+  // reduce/scan/dmapped are precedence 15, but don't come through this path.
+  else if (strcmp(op, "!") == 0 || strcmp(op, "~") == 0)
+    return 14;
+  else if (strcmp(op, "*") == 0 ||
+           strcmp(op, "/") == 0 || strcmp(op, "%") == 0)
+    return 13;
+  else if (unary &&
+           (strcmp(op, "+") == 0 || strcmp(op, "-") == 0))
+    return 12;
+  else if (strcmp(op, "<<") == 0 || strcmp(op, ">>") == 0)
+    return 11;
+  else if (strcmp(op, "&") == 0)
+    return 10;
+  else if (strcmp(op, "^") == 0)
+    return 9;
+  else if (strcmp(op, "|") == 0)
+    return 8;
+  else if (strcmp(op, "+") == 0 || strcmp(op, "-") == 0)
+    return 7;
+  // .. and ..< are precedence 6, but don't come through this path.
+  else if (strcmp(op, "<") == 0 || strcmp(op, "<=") == 0 ||
+           strcmp(op, ">") == 0 || strcmp(op, ">=") == 0)
+    return 5;
+  else if (strcmp(op, "==") == 0 || strcmp(op, "!=") == 0)
+    return 4;
+  else if (strcmp(op, "&&") == 0)
+    return 3;
+  else if (strcmp(op, "||") == 0)
+    return 2;
+  // by and align are precedence 1 too, but don't come through this path.
+  else if (strcmp(op, "#") == 0)
+    return 1;
+  printf("unknown op %s", op);
```

Comment:
Typically we handle missed cases by altering the generated docs output to include the `AppendExpr.Call01` or whatever output.  However, this is not a function that generates such output, this is a function that is used to compute something that alters the output and the computation is not accurate if a case is missed, right?  There's probably other ways to modify the function so that the error is passed to some place that handles it better

Reply:
Hmm.  It could return `true` - parentheses are needed -- for any operators it doesn't recognize.  That would be "safe" in that adding parens around the printout of an AST node won't be wrong, just perhaps unneeded.

Could it use a `USR_WARN()`?  I'm not able to envision what to print in the output, or where to print it, for something like `proc foo(x=bar(), y = a unknownop b, z=42)`

Reply:
USR_WARN is more for something the user has done wrong or that they should adjust.  But this seems more like something we need to adjust the compiler for when it is encountered, it's more our fault than the user's.

Unnecessary parentheses seems like a fine fallback to me - if the user doesn't like it, they can always open an issue about it.  I don't think that leads to a scenario where we notice if we've caused a case that has too many parentheses, though, but maybe that's okay?

Reply:
Ok, that makes sense.  I made that change.  (no change in any output since no test was hitting that INT_FATAL. :joy:  )